src/mscorlib/src/System/Text/EncodingForwarder.cs

```diff
@@ -179,19 +165,169 @@ public unsafe static int GetBytes(Encoding encoding, char[] chars, int charIndex
         public unsafe static int GetBytes(Encoding encoding, char* chars, int charCount, byte* bytes, int byteCount)
         {
             Debug.Assert(encoding != null);
-            if (bytes == null || chars == null)
+            // Validate parameters
+            if ((bytes == null) ||
+                (chars == null) ||
+                (charCount < 0) ||
+                (byteCount < 0))
             {
-                throw new ArgumentNullException(bytes == null ? nameof(bytes) : nameof(chars), Environment.GetResourceString("ArgumentNull_Array"));
-            }
-            if (charCount < 0 || byteCount < 0)
-            {
-                throw new ArgumentOutOfRangeException(charCount < 0 ? nameof(charCount) : nameof(byteCount), Environment.GetResourceString("ArgumentOutOfRange_NeedNonNegNum"));
+                ThrowValidationFailedException(chars, charCount, bytes);
             }
             Contract.EndContractBlock();
 
             return encoding.GetBytes(chars, charCount, bytes, byteCount, encoder: null);
         }
 
+        internal unsafe static bool TryEncode(char* input, int charCount, byte* output, int byteCount, out int charactersConsumed, out int bytesWritten)
+        {
+            const int Shift16Shift24 = (1 << 16) | (1 << 24);
+            const int Shift8Identity = (1 <<  8) | (1);
+
+            int charsToEncode = Math.Min(charCount, byteCount);
+
+            // Encode as bytes upto the first non-ASCII byte and return count encoded
```

Comment:
I meant:
- Use the complete implementation of this method from https://github.com/dotnet/corefxlab/pull/1156. Its performance for the ASCII case is within 5% of this implementation (varies with alignment of various things), it is pretty much same as what `int GetBytes(char* chars, int charCount, byte* bytes, int byteCount, EncoderNLS baseEncoder)` in this file is today, and thus it handles everything - except for the EncodingFallback that kills the performance.
- Leave refactored handling of EncodingFallback `int GetBytes(char* chars, int charCount, byte* bytes, int byteCount, EncoderNLS baseEncoder)` and call TryEncode to do the actual work.